File: apps/astrabridge-sidecar/astrabridge_sidecar/agent_orchestration_sdk.py
from __future__ import annotations

import json
from copy import deepcopy
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Mapping, Sequence

from .agent_orchestration_compiler import compile_agent_orchestration_graph
from .agent_orchestration_contract import (
    AGENT_ORCHESTRATION_SCHEMA_VERSION,
    lower_agent_orchestration_graph_to_task_graph,
    validate_agent_orchestration_graph,
)


DEFAULT_METADATA_CREATED_AT = "2026-07-07T00:00:00+09:00"
DEFAULT_METADATA_UPDATED_AT = "2026-07-07T00:05:00+09:00"
DEFAULT_COMPILED_TASK_GRAPH_VERSION = "astrabridge-task-graph-v1"
DEFAULT_COMPATIBILITY_NOTES = (
    "Canonical graphs remain the source of truth for GUI, code, dry-run, and runtime work.",
    "Lowering into legacy task graphs is a compatibility shim while the generic scheduler is still under construction.",
)


def _deepcopy_mapping(value: Mapping[str, Any] | None) -> dict[str, Any]:
    return deepcopy(dict(value or {}))


@dataclass(slots=True)
class CapabilityClaimsSpec:
    input_port_types: Sequence[str] = ()
    output_port_types: Sequence[str] = ()

    def to_payload(self) -> dict[str, Any]:
        return {
            "input_port_types": [str(item) for item in self.input_port_types],
            "output_port_types": [str(item) for item in self.output_port_types],
        }


@dataclass(slots=True)
class RoutingSpec:
    selection_mode: str = "none"
    provider_id: str | None = None
    model_id: str | None = None
    profile_id: str | None = None
    capability_claims: CapabilityClaimsSpec | None = None

    def to_payload(self) -> dict[str, Any]:
        payload: dict[str, Any] = {"selection_mode": self.selection_mode}
        if self.provider_id:
            payload["provider_id"] = self.provider_id
        if self.model_id:
            payload["model_id"] = self.model_id
        if self.profile_id:
            payload["profile_id"] = self.profile_id
        if self.capability_claims is not None:
            payload["capability_claims"] = self.capability_claims.to_payload()
        return payload


@dataclass(slots=True)
class PromptSpec:
    template: str
    template_mode: str = "inline"

    def to_payload(self) -> dict[str, Any]:
        return {
            "template_mode": self.template_mode,
            "template": self.template,
        }


@dataclass(slots=True)
class ToolsSpec:
    approval_mode: str = "ask"
    allowed_tool_classes: Sequence[str] = ()

    def to_payload(self) -> dict[str, Any]:
        return {
            "approval_mode": self.approval_mode,
            "allowed_tool_classes": [str(item) for item in self.allowed_tool_classes],
        }


@dataclass(slots=True)
class PortSpec:
    port_id: str
    label: str
    port_type: str
    shape: str = "single"
    required: bool = True
    schema_ref: str | None = None
    artifact_kind: str | None = None

    def to_payload(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "port_id": self.port_id,
            "label": self.label,
            "port_type": self.port_type,
            "shape": self.shape,
            "required": self.required,
        }
        if self.schema_ref is not None:
            payload["schema_ref"] = self.schema_ref
        if self.artifact_kind is not None:
            payload["artifact_kind"] = self.artifact_kind
        return payload


@dataclass(slots=True)
class InputContractSpec:
    mode: str
    port_ids: Sequence[str]

    def to_payload(self) -> dict[str, Any]:
        return {
            "mode": self.mode,
            "port_ids": [str(item) for item in self.port_ids],
        }


@dataclass(slots=True)
class ArtifactSpec:
    kind: str
    id: str

    def to_payload(self) -> dict[str, Any]:
        return {
            "kind": self.kind,
            "id": self.id,
        }


@dataclass(slots=True)
class OutputContractSpec:
    mode: str
    machine_result_schema_ref: str | None = None
    artifact_specs: Sequence[ArtifactSpec] = ()
    human_summary_required: bool = True

    def to_payload(self) -> dict[str, Any]:
        return {
            "mode": self.mode,
            "machine_result_schema_ref": self.machine_result_schema_ref,
            "artifact_specs": [item.to_payload() for item in self.artifact_specs],
            "human_summary_required": self.human_summary_required,
        }


@dataclass(slots=True)
class RetryPolicySpec:
    max_attempts: int = 1

    def to_payload(self) -> dict[str, Any]:
        return {"max_attempts": self.max_attempts}


@dataclass(slots=True)
class SubagentPolicySpec:
    isolation_mode: str = "lane"
    max_turns: int = 8
    allow_direct_teammate_messages: bool = False
    share_worktree: bool = False
    allow_nested_subagents: bool = False

    def to_payload(self) -> dict[str, Any]:
        return {
            "isolation_mode": self.isolation_mode,
            "max_turns": self.max_turns,
            "allow_direct_teammate_messages": self.allow_direct_teammate_messages,
            "share_worktree": self.share_worktree,
            "allow_nested_subagents": self.allow_nested_subagents,
        }


@dataclass(slots=True)
class ExecutionSpec:
    spawn_mode: str
    timeout_ms: int
    retry_policy: RetryPolicySpec = field(default_factory=RetryPolicySpec)
    execution_backend: str = "app_server"
    collaboration_mode: str = "default"
    subagent_policy: SubagentPolicySpec | None = None

    def to_payload(self) -> dict[str, Any]:
        return {
            "spawn_mode": self.spawn_mode,
            "timeout_ms": self.timeout_ms,
            "retry_policy": self.retry_policy.to_payload(),
            "execution_backend": self.execution_backend,
            "collaboration_mode": self.collaboration_mode,
            "subagent_policy": self.subagent_policy.to_payload() if self.subagent_policy is not None else None,
        }


@dataclass(slots=True)
class SafetySpec:
    risk_class: str
    allow_provider_calls: bool
    allow_code_changes: bool
    allow_install: bool
    requires_human_approval: bool
    approval_kind: str | None = None

    def to_payload(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "risk_class": self.risk_class,
            "allow_provider_calls": self.allow_provider_calls,
            "allow_code_changes": self.allow_code_changes,
            "allow_install": self.allow_install,
            "requires_human_approval": self.requires_human_approval,
        }
        if self.approval_kind is not None:
            payload["approval_kind"] = self.approval_kind
        return payload


@dataclass(slots=True)
class PositionSpec:
    x: int
    y: int

    def to_payload(self) -> dict[str, int]:
        return {"x": self.x, "y": self.y}


@dataclass(slots=True)
class UiSpec:
    position: PositionSpec
    layout_mode: str = "canvas"

    def to_payload(self) -> dict[str, Any]:
        return {
            "position": self.position.to_payload(),
            "layout_mode": self.layout_mode,
        }


@dataclass(slots=True)
class NodeSpec:
    node_id: str
    kind: str
    label: str
    role: str
    card_ref: str
    routing: RoutingSpec
    prompt: PromptSpec
    tools: ToolsSpec
    inputs: Sequence[PortSpec]
    outputs: Sequence[PortSpec]
    input_contract: InputContractSpec
    output_contract: OutputContractSpec
    execution: ExecutionSpec
    safety: SafetySpec
    ui: UiSpec
    status: str = "ready"

    def to_payload(self) -> dict[str, Any]:
        return {
            "node_id": self.node_id,
            "kind": self.kind,
            "label": self.label,
            "role": self.role,
            "card_ref": self.card_ref,
            "routing": self.routing.to_payload(),
            "prompt": self.prompt.to_payload(),
            "tools": self.tools.to_payload(),
            "ports": {
                "inputs": [item.to_payload() for item in self.inputs],
                "outputs": [item.to_payload() for item in self.outputs],
            },
            "input_contract": self.input_contract.to_payload(),
            "output_contract": self.output_contract.to_payload(),
            "execution": self.execution.to_payload(),
            "safety": self.safety.to_payload(),
            "ui": self.ui.to_payload(),
            "status": self.status,
        }


@dataclass(slots=True)
class PortBindingSpec:
    from_port_id: str
    to_port_id: str

    def to_payload(self) -> dict[str, str]:
        return {
            "from_port_id": self.from_port_id,
            "to_port_id": self.to_port_id,
        }


@dataclass(slots=True)
class HandoffContractSpec:
    message_template: str
    required_output_schema_refs: Sequence[str]
    port_bindings: Sequence[PortBindingSpec]
    message_part_modes: Sequence[str] = ("machine_result", "human_summary")

    def to_payload(self) -> dict[str, Any]:
        return {
            "message_template": self.message_template,
            "message_part_modes": [str(item) for item in self.message_part_modes],
            "required_output_schema_refs": [str(item) for item in self.required_output_schema_refs],
            "port_bindings": [item.to_payload() for item in self.port_bindings],
        }


@dataclass(slots=True)
class ContextPolicySpec:
    policy_id: str
    history_mode: str = "latest_summary_only"
    artifact_mode: str = "required_output_only"
    exclude_private_memory: bool = True
    include_machine_results: bool = True
    include_human_summaries: bool = True
    summary_strategy: str = "human_and_machine"
    history_length: int | None = None
    resource_refs: Sequence[str] = ()
    included_artifacts: Sequence[str] = ()

    def to_payload(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "policy_id": self.policy_id,
            "history_mode": self.history_mode,
            "artifact_mode": self.artifact_mode,
            "exclude_private_memory": self.exclude_private_memory,
            "include_machine_results": self.include_machine_results,
            "include_human_summaries": self.include_human_summaries,
            "summary_strategy": self.summary_strategy,
        }
        if self.history_length is not None:
            payload["history_length"] = self.history_length
        if self.resource_refs:
            payload["resource_refs"] = [str(item) for item in self.resource_refs]
        if self.included_artifacts:
            payload["included_artifacts"] = [str(item) for item in self.included_artifacts]
        return payload


@dataclass(slots=True)
class EdgeSpec:
    edge_id: str
    from_node_id: str
    to_node_id: str
    edge_type: str
    handoff_contract: HandoffContractSpec
    context_policy: ContextPolicySpec
    ui: UiSpec
    status: str = "ready"

    def to_payload(self) -> dict[str, Any]:
        return {
            "edge_id": self.edge_id,
            "from_node_id": self.from_node_id,
            "to_node_id": self.to_node_id,
            "edge_type": self.edge_type,
            "handoff_contract": self.handoff_contract.to_payload(),
            "context_policy": self.context_policy.to_payload(),
            "ui": self.ui.to_payload(),
            "status": self.status,
        }


@dataclass(slots=True)
class GraphMetadataSpec:
    description: str
    tags: Sequence[str] = ()
    owners: Sequence[str] = ()
    created_at: str = DEFAULT_METADATA_CREATED_AT
    updated_at: str = DEFAULT_METADATA_UPDATED_AT

    def to_payload(self) -> dict[str, Any]:
        return {
            "description": self.description,
            "tags": [str(item) for item in self.tags],
            "owners": [str(item) for item in self.owners],
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }


@dataclass(slots=True)
class GraphPolicySpec:
    entry_node_ids: Sequence[str]
    max_depth: int = 2
    default_permission_mode: str = "ask"
    default_collaboration_mode: str = "default"
    default_execution_backend: str = "app_server"
    requires_dry_run_before_live: bool = True

    def to_payload(self) -> dict[str, Any]:
        return {
            "entry_node_ids": [str(item) for item in self.entry_node_ids],
            "max_depth": self.max_depth,
            "default_permission_mode": self.default_permission_mode,
            "default_collaboration_mode": self.default_collaboration_mode,
            "default_execution_backend": self.default_execution_backend,
            "requires_dry_run_before_live": self.requires_dry_run_before_live,
        }


@dataclass(slots=True)
class CompatibilitySpec:
    lowering_mode: str = "lossy_legacy_task_graph"
    preserves_unknown_fields: bool = False
    notes: Sequence[str] = DEFAULT_COMPATIBILITY_NOTES

    def to_payload(self) -> dict[str, Any]:
        return {
            "lowering_mode": self.lowering_mode,
            "preserves_unknown_fields": self.preserves_unknown_fields,
            "notes": [str(item) for item in self.notes],
        }


@dataclass(slots=True)
class MigrationSpec:
    source_kind: str = "native_authoring"
    compiled_task_graph_version: str = DEFAULT_COMPILED_TASK_GRAPH_VERSION
    compatibility: CompatibilitySpec = field(default_factory=CompatibilitySpec)

    def to_payload(self) -> dict[str, Any]:
        return {
            "source_kind": self.source_kind,
            "compiled_task_graph_version": self.compiled_task_graph_version,
            "compatibility": self.compatibility.to_payload(),
        }


@dataclass(slots=True)
class AgentOrchestrationGraphBuilder:
    graph_id: str
    task_id: str
    title: str
    metadata: GraphMetadataSpec
    graph_policy: GraphPolicySpec
    template_id: str = ""
    status: str = "ready"
    state_version: int = 1
    migration: MigrationSpec = field(default_factory=MigrationSpec)
    prompt_registry: Mapping[str, Any] | None = None
    external_agent_card_registry: Mapping[str, Any] | None = None
    nodes: list[NodeSpec] = field(default_factory=list)
    edges: list[EdgeSpec] = field(default_factory=list)
    schema_registry: dict[str, Any] = field(default_factory=dict)

    def add_node(self, node: NodeSpec) -> AgentOrchestrationGraphBuilder:
        self.nodes.append(node)
        return self

    def add_edge(self, edge: EdgeSpec) -> AgentOrchestrationGraphBuilder:
        self.edges.append(edge)
        return self

    def register_schema(self, schema_ref: str, schema: Mapping[str, Any]) -> AgentOrchestrationGraphBuilder:
        self.schema_registry[str(schema_ref)] = _deepcopy_mapping(schema)
        return self

    def build(self) -> dict[str, Any]:
        payload = self._build_payload()
        validate_agent_orchestration_graph(payload)
        return payload

    def _build_payload(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "schema_version": AGENT_ORCHESTRATION_SCHEMA_VERSION,
            "graph_id": self.graph_id,
            "task_id": self.task_id,
            "title": self.title,
            "status": self.status,
            "metadata": self.metadata.to_payload(),
            "graph_policy": self.graph_policy.to_payload(),
            "nodes": [item.to_payload() for item in self.nodes],
            "edges": [item.to_payload() for item in self.edges],
            "schema_registry": _deepcopy_mapping(self.schema_registry),
            "migration": self.migration.to_payload(),
            "state_version": self.state_version,
        }
        if self.template_id:
            payload["template_id"] = self.template_id
        if self.prompt_registry:
            payload["prompt_registry"] = _deepcopy_mapping(self.prompt_registry)
        if self.external_agent_card_registry:
            payload["external_agent_card_registry"] = _deepcopy_mapping(self.external_agent_card_registry)
        return payload

    def compile(self, **kwargs: Any) -> dict[str, Any]:
        return compile_agent_orchestration_graph(self.build(), **kwargs)

    def lower_to_task_graph(self) -> dict[str, Any]:
        return lower_agent_orchestration_graph_to_task_graph(self.build())

    def to_json(self) -> str:
        return json.dumps(self.build(), ensure_ascii=False, indent=2) + "\n"

    def write_json(self, path: str | Path) -> Path:
        file_path = Path(path)
        if file_path.suffix.lower() != ".json":
            raise ValueError("Python SDK graph output path must use the .json extension.")
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_text(self.to_json(), encoding="utf-8")
        return file_path

File: apps/astrabridge-desktop/src/features/runtime/agentOrchestrationSdk.ts
import type { AgentOrchestrationGraph } from "../../types";

export const AGENT_ORCHESTRATION_GRAPH_SCHEMA_VERSION = "astrabridge-agent-orchestration-graph-v1";

export type CapabilityClaimsSpec = {
  input_port_types?: string[];
  output_port_types?: string[];
};

export type RoutingSpec = {
  selection_mode: string;
  provider_id?: string;
  model_id?: string;
  profile_id?: string;
  capability_claims?: CapabilityClaimsSpec;
};

export type PromptSpec = {
  template_mode: string;
  template: string;
};

export type ToolsSpec = {
  approval_mode: string;
  allowed_tool_classes: string[];
};

export type PortSpec = {
  port_id: string;
  label: string;
  port_type: string;
  shape: string;
  required: boolean;
  schema_ref?: string;
  artifact_kind?: string;
};

export type InputContractSpec = {
  mode: string;
  port_ids: string[];
};

export type ArtifactSpec = {
  kind: string;
  id: string;
};

export type OutputContractSpec = {
  mode: string;
  machine_result_schema_ref: string | null;
  artifact_specs: ArtifactSpec[];
  human_summary_required: boolean;
};

export type RetryPolicySpec = {
  max_attempts: number;
};

export type SubagentPolicySpec = {
  isolation_mode: string;
  max_turns: number;
  allow_direct_teammate_messages: boolean;
  share_worktree: boolean;
  allow_nested_subagents: boolean;
};

export type ExecutionSpec = {
  spawn_mode: string;
  timeout_ms: number;
  retry_policy: RetryPolicySpec;
  execution_backend: string;
  collaboration_mode: string;
  subagent_policy: SubagentPolicySpec | null;
};

export type SafetySpec = {
  risk_class: string;
  allow_provider_calls: boolean;
  allow_code_changes: boolean;
  allow_install: boolean;
  requires_human_approval: boolean;
  approval_kind?: string;
};

export type PositionSpec = {
  x: number;
  y: number;
};

export type UiSpec = {
  position: PositionSpec;
  layout_mode: string;
};

export type NodeSpec = {
  node_id: string;
  kind: string;
  label: string;
  role: string;
  card_ref: string;
  routing: RoutingSpec;
  prompt: PromptSpec;
  tools: ToolsSpec;
  inputs: PortSpec[];
  outputs: PortSpec[];
  input_contract: InputContractSpec;
  output_contract: OutputContractSpec;
  execution: ExecutionSpec;
  safety: SafetySpec;
  ui: UiSpec;
  status?: string;
};

export type PortBindingSpec = {
  from_port_id: string;
  to_port_id: string;
};

export type HandoffContractSpec = {
  message_template: string;
  message_part_modes: string[];
  required_output_schema_refs: string[];
  port_bindings: PortBindingSpec[];
};

export type ContextPolicySpec = {
  policy_id: string;
  history_mode: string;
  artifact_mode: string;
  exclude_private_memory: boolean;
  include_machine_results: boolean;
  include_human_summaries: boolean;
  summary_strategy: string;
  history_length?: number;
  resource_refs?: string[];
  included_artifacts?: string[];
};

export type EdgeSpec = {
  edge_id: string;
  from_node_id: string;
  to_node_id: string;
  edge_type: string;
  handoff_contract: HandoffContractSpec;
  context_policy: ContextPolicySpec;
  ui: UiSpec;
  status?: string;
};

export type GraphMetadataSpec = {
  description: string;
  tags: string[];
  owners: string[];
  created_at: string;
  updated_at: string;
};

export type GraphPolicySpec = {
  entry_node_ids: string[];
  max_depth: number;
  default_permission_mode: string;
  default_collaboration_mode: string;
  default_execution_backend: string;
  requires_dry_run_before_live: boolean;
};

export type CompatibilitySpec = {
  lowering_mode: string;
  preserves_unknown_fields: boolean;
  notes: string[];
};

export type MigrationSpec = {
  source_kind: string;
  compiled_task_graph_version: string;
  compatibility: CompatibilitySpec;
};

export type AgentOrchestrationGraphBuilderArgs = {
  graph_id: string;
  task_id: string;
  title: string;
  template_id?: string;
  status?: string;
  metadata: GraphMetadataSpec;
  graph_policy: GraphPolicySpec;
  migration: MigrationSpec;
  state_version?: number;
  prompt_registry?: Record<string, unknown>;
  external_agent_card_registry?: Record<string, unknown>;
};

function cloneRecord<T extends Record<string, unknown>>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

function buildRouting(routing: RoutingSpec) {
  const payload: Record<string, unknown> = {
    selection_mode: routing.selection_mode,
  };
  if (routing.provider_id) payload.provider_id = routing.provider_id;
  if (routing.model_id) payload.model_id = routing.model_id;
  if (routing.profile_id) payload.profile_id = routing.profile_id;
  if (routing.capability_claims) {
    payload.capability_claims = {
      input_port_types: [...(routing.capability_claims.input_port_types ?? [])],
      output_port_types: [...(routing.capability_claims.output_port_types ?? [])],
    };
  }
  return payload;
}

function buildPort(port: PortSpec) {
  const payload: Record<string, unknown> = {
    port_id: port.port_id,
    label: port.label,
    port_type: port.port_type,
    shape: port.shape,
    required: port.required,
  };
  if (port.schema_ref !== undefined) payload.schema_ref = port.schema_ref;
  if (port.artifact_kind !== undefined) payload.artifact_kind = port.artifact_kind;
  return payload;
}

function buildNode(node: NodeSpec) {
  return {
    node_id: node.node_id,
    kind: node.kind,
    label: node.label,
    role: node.role,
    card_ref: node.card_ref,
    routing: buildRouting(node.routing),
    prompt: {
      template_mode: node.prompt.template_mode,
      template: node.prompt.template,
    },
    tools: {
      approval_mode: node.tools.approval_mode,
      allowed_tool_classes: [...node.tools.allowed_tool_classes],
    },
    ports: {
      inputs: node.inputs.map(buildPort),
      outputs: node.outputs.map(buildPort),
    },
    input_contract: {
      mode: node.input_contract.mode,
      port_ids: [...node.input_contract.port_ids],
    },
    output_contract: {
      mode: node.output_contract.mode,
      machine_result_schema_ref: node.output_contract.machine_result_schema_ref,
      artifact_specs: node.output_contract.artifact_specs.map((item) => ({
        kind: item.kind,
        id: item.id,
      })),
      human_summary_required: node.output_contract.human_summary_required,
    },
    execution: {
      spawn_mode: node.execution.spawn_mode,
      timeout_ms: node.execution.timeout_ms,
      retry_policy: {
        max_attempts: node.execution.retry_policy.max_attempts,
      },
      execution_backend: node.execution.execution_backend,
      collaboration_mode: node.execution.collaboration_mode,
      subagent_policy: node.execution.subagent_policy
        ? {
            isolation_mode: node.execution.subagent_policy.isolation_mode,
            max_turns: node.execution.subagent_policy.max_turns,
            allow_direct_teammate_messages: node.execution.subagent_policy.allow_direct_teammate_messages,
            share_worktree: node.execution.subagent_policy.share_worktree,
            allow_nested_subagents: node.execution.subagent_policy.allow_nested_subagents,
          }
        : null,
    },
    safety: {
      risk_class: node.safety.risk_class,
      allow_provider_calls: node.safety.allow_provider_calls,
      allow_code_changes: node.safety.allow_code_changes,
      allow_install: node.safety.allow_install,
      requires_human_approval: node.safety.requires_human_approval,
      ...(node.safety.approval_kind ? { approval_kind: node.safety.approval_kind } : {}),
    },
    ui: {
      position: {
        x: node.ui.position.x,
        y: node.ui.position.y,
      },
      layout_mode: node.ui.layout_mode,
    },
    status: node.status ?? "ready",
  };
}

function buildEdge(edge: EdgeSpec) {
  const contextPolicy: Record<string, unknown> = {
    policy_id: edge.context_policy.policy_id,
    history_mode: edge.context_policy.history_mode,
    artifact_mode: edge.context_policy.artifact_mode,
    exclude_private_memory: edge.context_policy.exclude_private_memory,
    include_machine_results: edge.context_policy.include_machine_results,
    include_human_summaries: edge.context_policy.include_human_summaries,
    summary_strategy: edge.context_policy.summary_strategy,
  };
  if (edge.context_policy.history_length !== undefined) {
    contextPolicy.history_length = edge.context_policy.history_length;
  }
  if (edge.context_policy.resource_refs?.length) {
    contextPolicy.resource_refs = [...edge.context_policy.resource_refs];
  }
  if (edge.context_policy.included_artifacts?.length) {
    contextPolicy.included_artifacts = [...edge.context_policy.included_artifacts];
  }
  return {
    edge_id: edge.edge_id,
    from_node_id: edge.from_node_id,
    to_node_id: edge.to_node_id,
    edge_type: edge.edge_type,
    handoff_contract: {
      message_template: edge.handoff_contract.message_template,
      message_part_modes: [...edge.handoff_contract.message_part_modes],
      required_output_schema_refs: [...edge.handoff_contract.required_output_schema_refs],
      port_bindings: edge.handoff_contract.port_bindings.map((item) => ({
        from_port_id: item.from_port_id,
        to_port_id: item.to_port_id,
      })),
    },
    context_policy: contextPolicy,
    ui: {
      position: {
        x: edge.ui.position.x,
        y: edge.ui.position.y,
      },
      layout_mode: edge.ui.layout_mode,
    },
    status: edge.status ?? "ready",
  };
}

export class AgentOrchestrationGraphBuilder {
  private readonly args: AgentOrchestrationGraphBuilderArgs;
  private readonly nodes: NodeSpec[] = [];
  private readonly edges: EdgeSpec[] = [];
  private readonly schemaRegistry: Record<string, unknown> = {};

  constructor(args: AgentOrchestrationGraphBuilderArgs) {
    this.args = args;
  }

  addNode(node: NodeSpec) {
    this.nodes.push(node);
    return this;
  }

  addEdge(edge: EdgeSpec) {
    this.edges.push(edge);
    return this;
  }

  registerSchema(schemaRef: string, schema: Record<string, unknown>) {
    this.schemaRegistry[schemaRef] = cloneRecord(schema);
    return this;
  }

  build(): AgentOrchestrationGraph {
    const graph = {
      schema_version: AGENT_ORCHESTRATION_GRAPH_SCHEMA_VERSION,
      graph_id: this.args.graph_id,
      task_id: this.args.task_id,
      title: this.args.title,
      ...(this.args.template_id ? { template_id: this.args.template_id } : {}),
      status: this.args.status ?? "ready",
      metadata: {
        description: this.args.metadata.description,
        tags: [...this.args.metadata.tags],
        owners: [...this.args.metadata.owners],
        created_at: this.args.metadata.created_at,
        updated_at: this.args.metadata.updated_at,
      },
      graph_policy: {
        entry_node_ids: [...this.args.graph_policy.entry_node_ids],
        max_depth: this.args.graph_policy.max_depth,
        default_permission_mode: this.args.graph_policy.default_permission_mode,
        default_collaboration_mode: this.args.graph_policy.default_collaboration_mode,
        default_execution_backend: this.args.graph_policy.default_execution_backend,
        requires_dry_run_before_live: this.args.graph_policy.requires_dry_run_before_live,
      },
      nodes: this.nodes.map(buildNode),
      edges: this.edges.map(buildEdge),
      schema_registry: cloneRecord(this.schemaRegistry),
      ...(this.args.prompt_registry ? { prompt_registry: cloneRecord(this.args.prompt_registry) } : {}),
      ...(this.args.external_agent_card_registry
        ? { external_agent_card_registry: cloneRecord(this.args.external_agent_card_registry) }
        : {}),
      migration: {
        source_kind: this.args.migration.source_kind,
        compiled_task_graph_version: this.args.migration.compiled_task_graph_version,
        compatibility: {
          lowering_mode: this.args.migration.compatibility.lowering_mode,
          preserves_unknown_fields: this.args.migration.compatibility.preserves_unknown_fields,
          notes: [...this.args.migration.compatibility.notes],
        },
      },
      state_version: this.args.state_version ?? 1,
    } satisfies AgentOrchestrationGraph;
    return graph;
  }

  toJson() {
    return `${JSON.stringify(this.build(), null, 2)}\n`;
  }
}

export function buildCustomBlankGraphFixture(taskId = "task_example") {
  const builder = new AgentOrchestrationGraphBuilder({
    graph_id: "graph_custom_blank_graph_v1",
    task_id: taskId,
    title: "Custom Blank Graph",
    template_id: "custom_blank_graph",
    metadata: {
      description: "Custom Blank Graph",
      tags: ["custom", "blank", "starter"],
      owners: [],
      created_at: "2026-07-07T00:00:00+09:00",
      updated_at: "2026-07-07T00:05:00+09:00",
    },
    graph_policy: {
      entry_node_ids: ["node_start_here"],
      max_depth: 2,
      default_permission_mode: "ask",
      default_collaboration_mode: "default",
      default_execution_backend: "app_server",
      requires_dry_run_before_live: true,
    },
    migration: {
      source_kind: "native_authoring",
      compiled_task_graph_version: "astrabridge-task-graph-v1",
      compatibility: {
        lowering_mode: "lossy_legacy_task_graph",
        preserves_unknown_fields: false,
        notes: [
          "Canonical graphs remain the source of truth for GUI, code, dry-run, and runtime work.",
          "Lowering into legacy task graphs is a compatibility shim while the generic scheduler is still under construction.",
        ],
      },
    },
  });

  builder.registerSchema("schema.blank_entry", {
    type: "object",
    required: ["goal", "next_nodes"],
  });

  builder.addNode({
    node_id: "node_start_here",
    kind: "artifact_source",
    label: "Start Here",
    role: "custom",
    card_ref: "agent_card_blank_entry",
    routing: {
      selection_mode: "none",
    },
    prompt: {
      template_mode: "inline",
      template: "Use this starter node as the seed for a custom graph.",
    },
    tools: {
      approval_mode: "ask",
      allowed_tool_classes: [],
    },
    inputs: [
      {
        port_id: "task_context",
        label: "Task Context",
        port_type: "text",
        shape: "single",
        required: true,
      },
    ],
    outputs: [
      {
        port_id: "machine_result",
        label: "Machine Result",
        port_type: "structured_json",
        shape: "single",
        required: true,
        schema_ref: "schema.blank_entry",
      },
    ],
    input_contract: {
      mode: "task_context_and_typed_ports",
      port_ids: ["task_context"],
    },
    output_contract: {
      mode: "structured_only",
      machine_result_schema_ref: "schema.blank_entry",
      artifact_specs: [],
      human_summary_required: true,
    },
    execution: {
      spawn_mode: "inline_lane",
      timeout_ms: 60000,
      retry_policy: {
        max_attempts: 1,
      },
      execution_backend: "app_server",
      collaboration_mode: "default",
      subagent_policy: null,
    },
    safety: {
      risk_class: "low",
      allow_provider_calls: false,
      allow_code_changes: false,
      allow_install: false,
      requires_human_approval: false,
    },
    ui: {
      position: {
        x: 140,
        y: 200,
      },
      layout_mode: "canvas",
    },
  });

  return builder;
}
